Capt 71 reuse existing http connection

paulb79 opened the pull request:
CAPT-55 Updated Traefik to asynchronously send audit events to Datastream

Implements local queue
Non blocking front end
May be configured with or without certificate for HTTP or HTTPS
Reuses HTTP connection
Updated LDS detection script with help from Phil Evans

Commit:
CAPT-45  add http sink

Commit:
CAPT-45 rename log attrib. key from msg to message

Commit:
CAPT-45 update log event prefix

Commit:
CAPT-45 disable security alert for dummy secret

paulb79 committed:
WIP FIXME

paulb79 committed:
Merge remote-tracking branch 'origin/CAPT-45_route-traefik-to-datastream2' into CAPT-45_route-traefik-to-datastream2

Commit:
CAPT-45 use mdtp.pem in http client

paulb79 committed:
Debug info

paulb79 committed:
Add logging

paulb79 committed:
Debug info

paulb79 committed:
Revert changing to lock to hopefully restore build

Commit:
CAPT-45 remove debug logging

Commit:
CAPT-68 add test ensuring events are logged on 502

Commit:
CAPT-68 add test to ensure http client is async

paulb79 committed:
Get the SSL Certificate from an environment variable

Ensure the connection is closed so that it may be reused

WIP fixing and updating version of Go to provide access to http.NewRequestWithContext

CAPT-71 reuse http connection

Configure the HTTP connection to ensure the connection is reused.

Provide test for reusing connections. Limitation of both Go 1.12 and my knowledge means this is not feasible to test so a parallel test is provide which mimics the behaviour of the Audit function in httpSink.go.

It is limited by the fact the types for client and clienttrace are different and therefore a logging http client can't be offered to Audit under test conditions.

Instead this test is verifying that http connections can be reused.

CAPT-71 reuse http connection

Configure the HTTP connection to ensure the connection is reused.

Provide test for reusing connections. Limitation of both Go 1.12 and my knowledge means this is not feasible to test so a parallel test is provide which mimics the behaviour of the Audit function in httpSink.go.

It is limited by the fact the types for client and clienttrace are different and therefore a logging http client can't be offered to Audit under test conditions.

Instead this test is verifying that http connections can be reused.

CAPT-71 include test helpers and test certs

CAPT-71 include test helpers and test certs

Commit:
CAPT-68 make http client async using on-disk queue

paulb79 committed:
Merge pull request #21 from hmrc/CAPT-68_add-tests

CAPT-45 add async http sink

Commit:
CAPT-68 remove log.Info

Commit:
CAPT-68 mimic formatting of last successful build

Commit:
CAPT-68 revert Gopkg.lock changes

paulb79 committed:
Fix leak detection

paulb79 committed:
Resolve issues with leak detection trips for test keys. And update to remove no longer required rules

paulb79 committed:
CAPT-71
Resolved testing issues.
1. Test for connection reuse
2. Resolves issue in sendRequest, if an error was found it wasn't propergated and so caused a SIGSEGV fault.
3. Added test to ensure http client can be created with a certficate path or not

Removed the httpSink as the async version is the only one required.

hughfdjackson commented on middlewares/audittap/streams/httpSinkAsync.go:
```diff
@@ -0,0 +1,204 @@
+	clientID := config.ClientID
+	if clientID == "" {
+		clientID = "hmrc-traefik-" + config.ProxyingFor
+	}
```
I might be mis-reading this code, but it seems like this isn't used?

paulb79 replied:
Done; removed

hughfdjackson commented on middlewares/audittap/streams/httpSinkAsync.go:
```diff
@@ -0,0 +1,204 @@
+	clientVersion := config.ClientVersion
+	if clientVersion == "" {
+		clientVersion = "not-set"
+	}
```
I might be mis-reading this code, but it seems like this isn't used?

paulb79 replied:
Done; removed

hughfdjackson commented on middlewares/audittap/streams/httpSinkAsync.go:
```diff
@@ -0,0 +1,204 @@
+	if len(certPath) > 0 {
+		caCert, err := ioutil.ReadFile(certPath)
+		if err != nil {
+			log.Info("Error Cert Read ", err)
```
Worth an 'error' level log?  And perhaps even a programme abort?

If a certificate exists as an env-var, but is malformed or inaccessible to traefik, is harsh, early feedback better?

paulb79 replied:
Done

hughfdjackson commented on middlewares/audittap/streams/httpSinkAsync.go:
```diff
@@ -0,0 +1,204 @@
+		log.Warn("DS_EventMissed_AuditFailureResponse audit item : " + string(encoded.Bytes))
+		return
+	}
+	defer res.Body.Close()
```
Does it make sense to hoist this to the second line of the function?  

I'm looking at [this](https://go.dev/tour/flowcontrol/12):

>  A defer statement defers the execution of a function until the surrounding function returns.
> The deferred call's arguments are evaluated immediately, but the function call is not executed until the surrounding function returns. 

So if you get an error, the `defer res.Body.Close()` is never hit, and therefore never executes at the function return?

paulb79 replied:
Done

hughfdjackson commented on middlewares/audittap/streams/httpSinkAsync.go:
```diff
@@ -0,0 +1,204 @@
+			FieldMap: log.FieldMap{
+				log.FieldKeyMsg: "message",
+			},
+		})
```
Is this somehow 'scoped' locally to this function?  If not, it might be worth putting this at the start of the program.

Otherwise, our first messages will come through with 'msg' until the first request is sent, then it'll switch to 'message'?

paulb79 replied:
OK. I've moved this to init()

hughfdjackson commented on middlewares/audittap/streams/httpSinkAsync_test.go:
```diff
@@ -0,0 +1,141 @@
+	messages := make(chan atypes.Encoded, 1)
+	w1, _ := NewHTTPSinkAsync(&config, messages)
+	_ = w1.Audit(encodedJSONSample)
+	time.Sleep(1000 * time.Millisecond)
```
ouch!  I realise that's probably unavailable, though.

hughfdjackson commented on middlewares/audittap/streams/httpSinkAsync_test.go:
```diff
@@ -0,0 +1,141 @@
+	log.SetOutput(&buf)
+
+	var sleepTime time.Duration
+	sleepTime = 2000 * time.Millisecond
```
Inline this?   It feels neater to me to have the declaration + initialisation together.  Don't have to worried about the variable being `nil` then.

paulb79 replied:
Done

paulb79 committed:
CAPT-71
Resolved testing issues.
1. Test for connection reuse
2. Resolves issue in sendRequest, if an error was found it wasn't propergated and so caused a SIGSEGV fault.
3. Added test to ensure http client can be created with a certficate path or not

Removed the httpSink as the async version is the only one required.
